lib/Runtime/Library/ScriptFunction.cpp

```diff
@@ -474,12 +363,22 @@ using namespace Js;
 
             charcount_t cch = pFuncBody->LengthInChars();
             size_t cbLength = pFuncBody->LengthInBytes();
-            LPCUTF8 pbStart = pFuncBody->GetSource(_u("ScriptFunction::EnsureSourceString"));
-            BufferStringBuilder builder(cch, scriptContext);
+            LPCUTF8 pbStart = pFuncBody->GetToStringSource(_u("ScriptFunction::EnsureSourceString"));
+            // cch and cbLength refer to the length of the parse, which may be smaller than the length of the to-string function
+            Assert(pFuncBody->StartOffset() >= pFuncBody->PrintableStartOffset());
+            size_t cbPreludeLength = pFuncBody->StartOffset() - pFuncBody->PrintableStartOffset();
+            Assert(cbPreludeLength < MaxCharCount);
+            // the toString of a function may include some prelude, e.g. the computed name expression.
+            // We do not store the char-index of the start, but if there are cbPreludeLength bytes difference,
+            // then that is an upper bound on the number of characters difference.
+            // We also assume that function.toString is relatively infrequent, and non-ascii characters in
+            // a prelude are relatively infrequent, so the inaccuracy here should in general be insignificant
+
+            BufferStringBuilder builder(cch + static_cast<charcount_t>(cbPreludeLength), scriptContext);
             utf8::DecodeOptions options = pFuncBody->GetUtf8SourceInfo()->IsCesu8() ? utf8::doAllowThreeByteSurrogates : utf8::doDefault;
-            size_t decodedCount = utf8::DecodeUnitsInto(builder.DangerousGetWritableBuffer(), pbStart, pbStart + cbLength, options);
+            size_t decodedCount = utf8::DecodeUnitsInto(builder.DangerousGetWritableBuffer(), pbStart, pbStart + cbLength + cbPreludeLength, options);
 
-            if (decodedCount != cch)
+            if (decodedCount < cch)
```

Comment:
I need to fix this up a little actually; when we overestimate we still need to go and update the length. May as well copy into a new properly allocated string in that case.

Reply:
Ok, that should be sorted now, although I'd appreciate a second set of eyes on this.

Reply:
Just out of curiosity, what’s the case in which we’re wrong about the length?

Reply:
Because I'm only tracking one number, I only know the byte-offset of the start of the "To-string" part of the source, I don't know the character-offset. This means that if there are characters outside of 7 bit ascii, then we don't have "1-byte = 1 character" any more, so we will allocate too many characters' worth of space.

For example in the test I've added I use the ß character which is 2 bytes in utf8. If it appears somewhere like `async /* ß */ foo() {}` as part of a member function, the bounds that we track fully are for parsing, and so are for `foo() {}` or just `() {}`, I'd have to go double check again to say which exactly. We do have a byte-offset to the start of the `async` and so we can fully recompute the string, but there are ~15 byte difference between the byte indexes, and only ~14 characters in that region (again, because the ß is 2 bytes for one character).